### PR TITLE
KAFKA-6813: Remove deprecated APIs in KIP-182, Part I

### PR DESCRIPTION
**Pull request by guozhangwang**
I'm breaking KAFKA-6813 into a couple of "smaller" PRs and this is the first one. It focused on:

1. Remove deprecated APIs in KStream, KTable, KGroupedStream, KGroupedTable, SessionWindowedKStream, TimeWindowedKStream.

2. Also found a couple of overlooked bugs while working on them:

2.a) In KTable.filter / mapValues without the additional parameter indicating the materialized stores, originally we will not materialize the store. After KIP-182 we mistakenly diverge the semantics: for KTable.mapValues it is still the case, for KTable.filter we will always materialize.

2.b) In XXStream/Table.reduce/count, we used to try to reuse the serdes since their types are pre-known (for `reduce` it is the same types for both key / value, for `count` it is the same types for key, and `Long` for value). This was somehow lost in the past refactoring.

2.c) We are enforcing to cast a `Serde<V>` to `Serde<VR>` for XXStream / Table.aggregate, for which the returned value type is NOT known, such the enforced casting should not be applied and we should require users to provide us the value serde if they believe the default ones are not applicable.

2.d) Whenever we are creating a new `MaterializedInternal` we are effectively incrementing the suffix index for the store / processor-node names. However in some places this `MaterializedInternal` is only used for validation, so the resulted processor-node / store suffix is not monotonic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java [hunk 1]
```diff
@@ -39,13 +39,21 @@ public Long apply() {
...
     };
+
     final Aggregator<K, V, Long> countAggregator = new Aggregator<K, V, Long>() {
         @Override
         public Long apply(K aggKey, V value, Long aggregate) {
             return aggregate + 1;
         }
     };
 
+    final Initializer<V> reduceInitializer = new Initializer<V>() {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:50]: This is moved from TimeWindowedStreamImpl, just to be consistent with the other const functions.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java [hunk 2]
```diff
@@ -109,69 +75,34 @@ private void determineIsQueryable(final String queryableStoreName) {
         Objects.requireNonNull(reducer, "reducer can't be null");
         Objects.requireNonNull(materialized, "materialized can't be null");
         final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materializedInternal
-                = new MaterializedInternal<>(materialized, builder, REDUCE_NAME);
+                = new MaterializedInternal<>(materialized, builder, AGGREGATE_NAME);
+        if (materializedInternal.keySerde() == null) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: This is for 2.b), ditto elsewhere.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 3]
```diff
@@ -132,57 +133,59 @@ public String apply(K key, V value) {
...
-                                name,
-                                processorSupplier,
-                                this.keySerde,
-                                this.valSerde,
-                                sourceNodes,
-                                builder.name(),
-                                true);
+        final String name = builder.newProcessorName(FILTER_NAME);
+
+        // only materialize if the state store is queryable
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: This is for 2.a), ditto for `mapValues`.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKStreamImpl.java [hunk 4]
```diff
@@ -73,35 +68,62 @@ public V apply() {
...
+    }
+
     @Override
     public <T> KTable<Windowed<K>, T> aggregate(final Initializer<T> initializer,
                                                 final Aggregator<? super K, ? super V, T> aggregator,
                                                 final Merger<? super K, T> sessionMerger) {
-        Objects.requireNonNull(initializer, "initializer can't be null");
-        Objects.requireNonNull(aggregator, "aggregator can't be null");
-        Objects.requireNonNull(sessionMerger, "sessionMerger can't be null");
-        return doAggregate(initializer, aggregator, sessionMerger, (Serde<T>) valSerde);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKStreamImpl.java:104]: This is for 2.c), ditto elsewhere.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java [hunk 5]
```diff
@@ -63,42 +63,33 @@
...
     @Override
     public KTable<Windowed<K>, Long> count(final Materialized<K, Long, WindowStore<Bytes, byte[]>> materialized) {
         Objects.requireNonNull(materialized, "materialized can't be null");
         final MaterializedInternal<K, Long, WindowStore<Bytes, byte[]>> materializedInternal
                 = new MaterializedInternal<>(materialized, builder, AGGREGATE_NAME);
+        if (materializedInternal.keySerde() == null) {
+            materializedInternal.withKeySerde(keySerde);
+        }
         if (materializedInternal.valueSerde() == null) {
-            materialized.withValueSerde(Serdes.Long());
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:78]: This is for 2.d), ditto elsewhere.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java [hunk 6]
```diff
@@ -63,42 +63,33 @@
...
-    @SuppressWarnings("unchecked")
-    private <VR> KTable<Windowed<K>, VR> doAggregate(final Initializer<VR> initializer,
-                                                     final Aggregator<? super K, ? super V, VR> aggregator,
-                                                     final Serde<VR> serde) {
-        final String storeName = builder.newStoreName(AGGREGATE_NAME);
-        return (KTable<Windowed<K>, VR>) aggregateBuilder.build(new KStreamWindowAggregate<>(windows, storeName, initializer, aggregator),
-                                                                AGGREGATE_NAME,
-                                                                windowStoreBuilder(storeName, serde),
-                                                                false);
+        return aggregate(initializer, aggregator, Materialized.<K, VR, WindowStore<Bytes, byte[]>>with(keySerde, null));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:92]: This is a meta explanation about the serde inheritance across multiple internal impl classes:

1. reduce: inherit the key and value serdes from the parent XXImpl class.
2. count: inherit the key serdes, enforce setting the `Serdes.Long()` for value serdes.
3. aggregate: inherit the key serdes, do not set for value serdes internally (line 92 here is for this case).

**Comment by guozhangwang** [comment]
@mjsax @bbejeck @vvcephei this PR is ready for reviews now.

File: streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java [hunk 7]
```diff
@@ -166,6 +166,8 @@ public void setup() {
                .groupBy(MockMapper.selectKeyKeyValueMapper())
                .count();
 
+        System.out.print(builder.build().describe());
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java]: Is this intentional?

**Review by bbejeck — APPROVED** [review]
Just one minor comment, otherwise LGTM.

**Comment by bbejeck** [comment]
Failure unrelated, exceded rate limit.

Retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java [hunk 8]
```diff
@@ -484,61 +172,42 @@
...
      * <pre>{@code
      * KafkaStreams streams = ... // compute sum
+     * String queryableStoreName = "storeName" // the queryableStoreName should be the name of the store as defined by the Materialized instance
      * ReadOnlyKeyValueStore<String,Long> localStore = streams.store(queryableStoreName, QueryableStoreTypes.<String, Long>keyValueStore());
      * String key = "some-key";
      * Long sumForKey = localStore.get(key); // key must be local (application state is shared over all running Kafka Streams instances)
      * }</pre>
      * For non-local keys, a custom RPC mechanism must be implemented using {@link KafkaStreams#allMetadata()} to
      * query the value of the key on a parallel running instance of your Kafka Streams application.
-     * <p>
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: why do we remove this?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java [hunk 9]
```diff
@@ -484,61 +172,42 @@
...
-     * Combine the value of records in this stream by the grouped key.
+     * Aggregate the values of records in this stream by the grouped key.
      * Records with {@code null} key or value are ignored.
-     * Combining implies that the type of the aggregate result is the same as the type of the input value
-     * (c.f. {@link #aggregate(Initializer, Aggregator, org.apache.kafka.streams.processor.StateStoreSupplier)}).
+     * Aggregating is a generalization of {@link #reduce(Reducer) combining via reduce(...)} as it, for example,
+     * allows the result to have a different type than the input values.
      * The result is written into a local {@link KeyValueStore} (which is basically an ever-updating materialized view)
-     * provided by the given {@code storeSupplier}.
+     * that can be queried using the provided {@code queryableStoreName}.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: this need to be updated?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java [hunk 10]
```diff
@@ -548,550 +217,39 @@
...
      * <p>
      * To query the local {@link KeyValueStore} it must be obtained via
-     * {@link KafkaStreams#store(String, QueryableStoreType) KafkaStreams#store(...)}.
-     * Use {@link org.apache.kafka.streams.processor.StateStoreSupplier#name()} to get the store name:
+     * {@link KafkaStreams#store(String, QueryableStoreType) KafkaStreams#store(...)}:
      * <pre>{@code
-     * KafkaStreams streams = ... // compute sum
-     * String queryableStoreName = storeSupplier.name();
+     * KafkaStreams streams = ... // some aggregation on value type double
      * ReadOnlyKeyValueStore<String,Long> localStore = streams.store(queryableStoreName, QueryableStoreTypes.<String, Long>keyValueStore());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: in the example above, 

```
String queryableStoreName = "storeName" // the queryableStoreName should be the name of the store as defined by the Materialized instance
```

was inserted. should we do the same here?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 11]
```diff
@@ -2072,7 +1256,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * a value (with arbitrary type) for the result record.
      * The key of the result record is the same as for both joining input records.
      * Furthermore, for each input record of both {@code KStream}s that does not satisfy the join predicate the provided
-     * {@link ValueJoiner} will be called with a {@code null} value for this/other stream, respectively.
+     * {@link ValueJoiner} will be called with a {@code null} value for the this/other stream, respectively.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: typo?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 12]
```diff
@@ -1690,7 +887,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
      * If this requirement is not met, Kafka Streams will automatically repartition the data, i.e., it will create an
      * internal repartitioning topic in Kafka and write and re-read the data via this topic before the actual join.
      * The repartitioning topic will be named "${applicationId}-XXX-repartition", where "applicationId" is
-     * user-specified in {@link StreamsConfig} via parameter
+     * user-specified in {@link  StreamsConfig} via parameter
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:897]: nit: double space

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java [hunk 13]
```diff
@@ -109,69 +75,34 @@ private void determineIsQueryable(final String queryableStoreName) {
         Objects.requireNonNull(reducer, "reducer can't be null");
         Objects.requireNonNull(materialized, "materialized can't be null");
         final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materializedInternal
-                = new MaterializedInternal<>(materialized, builder, REDUCE_NAME);
+                = new MaterializedInternal<>(materialized, builder, AGGREGATE_NAME);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: Why the change to `AGGREGATE_NAME` ?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 14]
```diff
@@ -107,57 +107,59 @@ public KTableImpl(final InternalStreamsBuilder builder,
...
+                    true);
+        } else {
+            processorSupplier = new KTableFilter<>(this,
+                    predicate,
+                    filterNot,
+                    null);
+
+            builder.internalTopologyBuilder.addProcessor(name, processorSupplier, this.name);
+
+            return new KTableImpl<>(builder,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: might be better to move the return after the if-then-else to unify code and just use a boolean variable for the last parameter?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 15]
```diff
@@ -107,57 +107,59 @@ public KTableImpl(final InternalStreamsBuilder builder,
...
-                                this.valSerde,
-                                sourceNodes,
-                                builder.name(),
-                                true);
+        final String name = builder.newProcessorName(FILTER_NAME);
+
+        // only materialize if the state store is queryable
+        KTableProcessorSupplier<K, V, V> processorSupplier;
+
+        if (materialized != null && materialized.isQueryable()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: As this is a `private` method, we could simplify be never calling with `null`?

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationDedupIntegrationTest.java [hunk 16]
```diff
@@ -58,7 +63,7 @@
 import static org.hamcrest.core.Is.is;
 
 /**
- * Similar to KStreamAggregationIntegrationTest but with dedupping enabled
+ * Similar to KStreamAggregationIntegrationTest but with de dupping enabled
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationDedupIntegrationTest.java]: nit: I think `deduping` is correct?

**Review by mjsax — COMMENTED** [review]
Pretty hard to review... Overall looks good. Couple of questions.

Also one meta comment about Serde inheritance: Isn't it inconsistent to do this for `groupBy-aggregate` only but not for other operators? For example a `builder.stream("", Consumed.with(...)).filter.to("", Produced.with())` could forward the Serdes from the source via the filter to the sink making the `Produced.with()` unnecessary. I like Serde inheritance, but should we do it consistently instead of picking a "random" pair of operators that support it?

File: streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java]: No :P Will remove it.

**Comment by guozhangwang** [comment]
@mjsax I understand this is a large PR... I've tried hard to make it into multiple smaller ones (I'll probably have two other PRs at about similar sizes) while the hope is that for this PR, most of the removals are straight forward except the fixes mentioned in 2).

Regarding your question about inheritance: I did this for the following reasons:

a) Before KIP-182 we are actually doing sth. already about the inheritance, but only for count and reduce; when we add KIP-182 we mistakenly dropped some of those.
b) For aggregations, i.e. from grouped stream / grouped table / time windowed grouped stream, the only following operation is aggregation; while for other operators like `stream().to()` we need to decide whether to inherit based on the actual operator, which is more tricky. I do have plans to enhance the inheritance of serdes once we have removed the deprecation though (that was part of the plan for post-KIP182 anyways) but I want separate that general enhancement from this smaller scoped feature.
c) Scala API is actually relying on the fix above to make strong typing strict. So I want to make sure this part does make it to 2.0 release along with the Scala API.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java (hunk 8 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: Hmm.. good point, I will add them back.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:897]: ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: This is a good one: while working on this class I realized that we mistakenly used `REDUCE_NAME` for the prefix, while before this we used `AGGREGATE_NAME`, and in `count` and other `reduce` places we use `AGGREGATE_NAME` as well. 

So technically I think this is the right fix, but arguably it will result in different store names if users are on this version already.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Ack. Will try to do the same for mapValues as well.

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationDedupIntegrationTest.java (hunk 16 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationDedupIntegrationTest.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 15 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: I thought about that, but this would require the other callers to pass in a dummy Materialized which I think does not worth the code cleanness. I've refactored the part according to your comment above and LMK what do you think.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java (hunk 13 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: > `we mistakenly used REDUCE_NAME for the prefix`

Why "mistakenly" -- it's a `reduce(...)` operator? Because `count()` is an special case of `aggregation()` it seems to be fine to use `AGGREGATE_NAME` there?

File: streams/src/test/java/org/apache/kafka/streams/TopologyWrapper.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+ */
+package org.apache.kafka.streams;
+
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+
+/**
+ *  This class allows to access the {@link InternalTopologyBuilder} a {@link Topology} object.
+ *
+ */
+public class TopologyWrapper extends Topology {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/TopologyWrapper.java]: Was this added on purpose in the last commit? Seems to be unused.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/TopologyWrapper.java]: Not on purpose, it was added for a follow-up PR. Will revert for now.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: There are three places where `reduce` were called: KGroupedStream, TimeWindowedKStream, and SessionWindowedKStream. What I've observed is that in the first two we use `REDUCE_NAME` as both processor name and store name prefixes, and in the latter we use `AGGREGTE_NAME` for both processor name and store name prefixes. I thought the latter was right and the former is not. But if people can correct me that it is the other way around I'm happy to change in the other way (either case, there is a compatibility concern).

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: I agree about the compatibility concern -- if we change the name, it might break the upgrade path. To me, `REDUCE_NAME` seems to be correct and thus, if we really change align all of them (not sure if it's worth it), we should update `SessionWindowedKStream#reduce()`.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: Sounds good, will leave a note on the upgrade path as well.

**Review by mjsax — APPROVED** [review]
Feel free to merge. Please don't forget to update the docs --  can be done in follow up PR, too.
